### PR TITLE
week13

### PR DESCRIPTION
**Pull request by ofeterovskaya**
create Server Side Rendering, create package json, install packages

### CODE REVIEW

File: app.js
Comment (mattcarlotta) [inline app.js]: Requires change:
Move this dotenv import to the top of the file. You'll want to ensure this loads first because some 3rd packages may expect a process.env to be defined before being imported.

File: app.js
Comment (mattcarlotta) [inline app.js]: Nitpick:
This logic can be simplified. For example, you can set the cookie to `secure` within the `sessionParms`:
```js
const inProduction = app.get("env") === "production";

const sessionParms = {
    secret: process.env.SESSION_SECRET,
    resave: true,
    saveUninitialized: true,
    store: store,
    cookie: { secure: false, sameSite: "strict", secure: inProduction },
};

if (inProduction) {
    app.set("trust proxy", 1); // trust first proxy
}
```

File: app.js
Comment (mattcarlotta) [inline app.js]: if (!req.session.secretWord) {
  req.session.secretWord = "syzygy";
}

File: app.js
Comment (mattcarlotta) [inline app.js]: Nitpick:
This logic seems a bit overly complicated, does the opposite of a more common approach to checking a string,  and will fail if the `req.body.secretWord` is undefined. Instead of uppercasing, you'll often see it lowercased; then, you can use [optional chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) and `startsWith`!
```diff
- if (req.body.secretWord.toUpperCase()[0] == "P") {
+ if (req.body.secretWord?.toLowerCase()?.startsWith("p")) {
```
Alternatively, you could use a RegEx, but I'm not a fan of them as they're kind of pain to manage.

File: views/partials/head.ejs
Comment (mattcarlotta) [inline views/partials/head.ejs:1]: Requires change:
Same as above, I'd highly recommend removing the `DOCTYPE`, `html` and `body` tags from this file because this ties and enforces the partial template to be used within a non-partial file to open and close tags. Instead make it stand-alone:
```html
<head>
    <meta charset="UTF-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <link href="https://cdn.jsdelivr.net/npm/bootstrap@5.2.2/dist/css/bootstrap.min.css" rel="stylesheet"
        integrity="sha384-Zenh87qX5JnK2Jl0vWa8Ck2rdkQ2Bzep5IDxbcnCeuOxjzrPF/et3URy9Bv1WTRi" crossorigin="anonymous" />
    <title>Secret Word</title>
</head>
```

📓 One could argue that these `head` tags should also be removed and added to the non-partial template. But for simplicity, I've kept them in this file.

File: views/partials/header.ejs
Comment (mattcarlotta) [inline views/partials/header.ejs:1]: Nitpick:
Update `h1` tag text to reflect the "Secret Word" project:
```html
<header>
    <h1 class="fs-1">The Secret Word EJS Application</h1>
</header>
```

File: views/partials/header.ejs
Comment (mattcarlotta) [inline views/partials/header.ejs]: Requires change:
These if conditions are always true because `errors` and `info` are initially empty arrays and empty arrays are defined and therefore `true`. Instead, you'll want to check if defined AND has length:
```ejs
<% if (errors && errors.length) {
    errors.forEach((err) => { %>
      <p class="alert alert-danger">Error: <%= err %></p>
<% }) } %>
<% if (info && info.length) {
    info.forEach((msg) => { %>
      <p class="alert alert-success">Info: <%= msg %></p>
<% }) } %>
```
🗒️ You're currently importing Bootstrap CSS within the `head.ejs` file, so utilize it! For example, leverage Bootstrap's [alerts](https://getbootstrap.com/docs/5.3/components/alerts/#examples).

File: views/partials/header.ejs
Comment (mattcarlotta) [inline views/partials/header.ejs]: Nitpick:
I'd recommend wrapping the `hr` tag in a conditional so that it doesn't render until there's a message:
```ejs
<% if ((errors && errors.length) || (info && info.length)) { %>
    <hr />
<% } %>
```

File: views/secretWord.ejs
Comment (mattcarlotta) [inline views/secretWord.ejs:1]: Requires change:
Since the partial templates are changed to stand-alone files, you can now require them within this non-partial file and not worry about opening/closing tags. This should simplify and make your non-partial templates reusable.
```ejs
<!DOCTYPE html>
<html lang="en">
    <%- include("partials/head.ejs") %>
    <body class="p-3">
        <%- include("partials/header.ejs") %>
        <main>
            <header>
                <h2 class="fs-4 mb-3">
                    The secret word is: 
                    <span class="badge text-bg-secondary fs-6">
                        <%= secretWord %>
                    </span>
                </h2>
            </header>
            <form method="POST">
                <div class="input-group mb-3">
                    <label class="input-group-text" for="secretWord">
                        Change Secret Word:&nbsp;
                    </label>
                    <input type="text" id="secretWord" name="secretWord" class="form-control" />
                 </div>
                 <button type="submit" class="btn btn-primary">
                     Submit
                 </button>
             </form>
         </main>
         <%- include("partials/footer.ejs") %>
    </body>
</html>
```
🗒️ You're currently importing Bootstrap CSS within the head.ejs file, so don't be afraid to leverage that Bootstrap CSS!

**Review by mattcarlotta — CHANGES_REQUESTED** [review]
Good work, Oksana! There's a few things that need to be tweaked, so please address ALL required feedback below! Once you've addressed the required changes, please request another review via GitHub or Slack.

File: app.js
Comment (ofeterovskaya) [reply app.js]: wow, I didn't think about this. Thank you for pointing this out

File: views/secretWord.ejs
Comment (ofeterovskaya) [reply views/secretWord.ejs:1]: I really enjoyed Bootstrap

File: app.js
Comment (mattcarlotta) [reply app.js]: Oops, this feedback is wrong! 😑 

Instead, I meant to write this…

Nitpick:
Here's a great example of how using the [Logical OR assignment](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Logical_OR_assignment) can simplify your code!
```diff
- if (!req.session.secretWord) {
-    req.session.secretWord = "syzygy";
- }
+ req.session.secretWord ||= "syzygy";
```

File: views/partials/footer.ejs
Comment (mattcarlotta) [inline views/partials/footer.ejs:1]: Requires change:
I'd highly recommend removing the ending `html` tag from this file because this ties and enforces the partial template to be used within a non-partial file to close HTML tag. Instead, I'd recommend that this partial template should be designed as a stand-alone template for a footer:
```ejs
<hr />
<footer>
    <p>&copy; <%= new Date().getFullYear() %> Oksana Feterovskaya</p>
</footer>
```

File: app.js
Comment (mattcarlotta) [inline app.js]: Oops, the previous feedback I left for this code is wrong! 😑 

Instead, I meant to write this…

Nitpick:
Here's a great example of how using the [Logical OR assignment](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Logical_OR_assignment) can simplify your code!
```diff
- if (!req.session.secretWord) {
-    req.session.secretWord = "syzygy";
- }
+ req.session.secretWord ||= "syzygy";
```

**Review by mattcarlotta — APPROVED** [review]
Excellent job, Oksana! Overall, the PR is looking good to me! That said, I realized I missed some feedback, and added it below, so please address it before merging! 👍

**Pull request merged**
